[JSC] `Intl.DurationFormat` should format seconds with `roundingMode` is `trunc`

sosukesuzuki opened the pull request:
#### 55784d54785e187216c433d10b4e0c504a722e98
<pre>
[JSC] `Intl.DurationFormat` should format seconds with `roundingMode` is `trunc`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275745">https://bugs.webkit.org/show_bug.cgi?id=275745</a>

Reviewed by Yusuke Suzuki.

According to the specifications [1][2], `Intl.DurationFormat` should use `trunc` as the
`roundingMode` when formatting seconds, milliseconds, or microseconds.

However, the current JSC is formatting with the default `roundingMode` of `Intl.NumberFormat`,
which is `halfExpand` (in practice, it uses the corresponding ICU rounding mode[3][4]
`rounding-mode-half-up`).

This patch changes the formatting of seconds, milliseconds, or microseconds to use the ICU rounding
mode `rounding-mode-down`, which corresponds to `trunc` as the `roundingMode`.

[1]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern">https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern</a>
[2]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-formatnumericseconds">https://tc39.es/proposal-intl-duration-format/#sec-formatnumericseconds</a>
[3]: <a href="https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/rounding-modes.md">https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/rounding-modes.md</a>
[4]: <a href="https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/skeletons.md#rounding-mode">https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/skeletons.md#rounding-mode</a>

* JSTests/stress/intl-durationformat-format-to-parts.js:
(Intl.DurationFormat.shouldBeOneOf):
* JSTests/stress/intl-durationformat-roundingMode.js: Added.
(sameValue):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::collectElements):

Canonical link: <a href="https://commits.webkit.org/281955@main">https://commits.webkit.org/281955@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac6d35371f2711f94d16c5241e13054d1ed922f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40933 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14166 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12124 "Build is in progress. Recent messages:") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12397 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/65553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/59/builds/12124 "Build is in progress. Recent messages:") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53333 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34688 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10565 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11055 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/54676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/56495 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10864 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60822 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5522 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/67282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53288 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4541 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82585 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9265 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14433 "Found 1 new JSC stress test failure: stress/sampling-profiler-display-name.js.dfg-eager-no-cjit-validate (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/37820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/38914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/test262/expectations.yaml:
```diff
@@ -1102,9 +1102,9 @@ test/intl402/DateTimeFormat/timezone-not-canonicalized.js:
-test/intl402/DurationFormat/prototype/format/rounding-mode-trunc-for-seconds.js:
-  default: 'Test262Error: Intl.DurationFormat should format seconds, milliseconds and microseconds with `roundingMode: "trunc"` Expected SameValue(«2», «1») to be true'
-  strict mode: 'Test262Error: Intl.DurationFormat should format seconds, milliseconds and microseconds with `roundingMode: "trunc"` Expected SameValue(«2», «1») to be true'
+test/intl402/DurationFormat/prototype/formatToParts/formatToParts-style-digital-en.js:
```
This change has introduced a new FAIL, which I believe is due to rounding errors from adding seconds as a double. In other words, I don't think this change is the direct cause. For example, the failure in the `precision-exact-mathematical-values.js` test is for the same reason.

To fundamentally solve the issue, we need to add seconds as an integer rather than as a double. For this, I am waiting for https://github.com/tc39/test262/pull/3988.

Constellation replied:
Sounds good.

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -359,13 +359,13 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
         case TemporalUnit::Second:
         case TemporalUnit::Millisecond:
         case TemporalUnit::Microsecond: {
+            skeletonBuilder.append("rounding-mode-down"_s);
```
Is this allowed in all ICU versions? Can you check when this is allowed, and ensure that it is enough old. (If not, you need to put runtime ICU version check).

sosukesuzuki replied:
`rounding-mode-down` is already used in [WebKit's `Intl.NumberFormat`](https://github.com/WebKit/WebKit/blob/bdc5e116475b8bd859d455fe8b37fd9d4cd07e42/Source/JavaScriptCore/runtime/IntlNumberFormatInlines.h#L198) without a runtime version check. So, I believed it would work without any issues in the ICU versions that WebKit intends to support.  However, should I still investigate the ICU version compatibility?

sosukesuzuki committed:
[JSC] `Intl.DurationFormat` should format seconds with `roundingMode` is `trunc`
https://bugs.webkit.org/show_bug.cgi?id=275745

Reviewed by Yusuke Suzuki.

According to the specifications [1][2], `Intl.DurationFormat` should use `trunc` as the
`roundingMode` when formatting seconds, milliseconds, or microseconds.

However, the current JSC is formatting with the default `roundingMode` of `Intl.NumberFormat`,
which is `halfExpand` (in practice, it uses the corresponding ICU rounding mode[3][4]
`rounding-mode-half-up`).

This patch changes the formatting of seconds, milliseconds, or microseconds to use the ICU rounding
mode `rounding-mode-down`, which corresponds to `trunc` as the `roundingMode`.

[1]: https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern
[2]: https://tc39.es/proposal-intl-duration-format/#sec-formatnumericseconds
[3]: https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/rounding-modes.md
[4]: https://github.com/unicode-org/icu/blob/main/docs/userguide/format_parse/numbers/skeletons.md#rounding-mode

* JSTests/stress/intl-durationformat-format-to-parts.js:
(Intl.DurationFormat.shouldBeOneOf):
* JSTests/stress/intl-durationformat-roundingMode.js: Added.
(sameValue):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::collectElements):

Canonical link: https://commits.webkit.org/281955@main
